src/whylogs/core/statistics/constraints.py

```diff
@@ -0,0 +1,272 @@
+
+
+class SummaryConstraint:
+    """
+    Summary constraints specify a relationship between a summary field and a literal value,
+    or between two summary fields.
+    e.g.     'min' < 6
+             'std_dev' < 2.17
+             'min' > 'avg'
+    """
+
+    def __init__(self, field: str, op: Op, value=None, field2: str = None, name: str = None, verbose=False):
+        self._verbose = verbose
+        self._name = name
+        self.op = op
+        self.field = field
+        self.field2 = field2
+        self.total = 0
+        self.failures = 0
+        if value is not None and field2 is None:
+            # field-value summary comparison
+            self.value = value
+            self.func = _summary_funcs1[op](field, value)
+        elif field2 is not None and value is None:
+            # field-field summary comparison
+            self.field2 = field2
+            self.func = _summary_funcs2[op](field, field2)
+        else:
+            raise ValueError("Summary constraint must specify a second value or field name, but not both")
+
+    @property
+    def name(self):
+        return self._name if self._name is not None else f'summary {self.field} {Op.Name(self.op)} {self.value}/{self.field2}'
+
+    def track(self, summ: NumberSummary) -> bool:
+        self.total += 1
+        if not self.func(summ):
+            self.failures += 1
+            if self._verbose:
+                logger.info(f'summary constraint {self.name} failed')
+
+    @staticmethod
+    def from_protobuf(msg: SummaryConstraintMsg) -> 'SummaryConstraint':
+        return ValueConstraint(msg.op, msg.value, name=msg.name)
+
+    def to_protobuf(self) -> SummaryConstraintMsg:
+        if self.field2 is None:
+            msg = SummaryConstraintMsg(
+                name=self.name,
+                field=self.field,
+                op=self.op,
+                value=self.value,
+            )
+        else:
+            msg = SummaryConstraintMsg(
+                name=self.name,
+                field=self.field,
+                op=self.op,
+                field2=self.field2,
+            )
+        return msg
+
+    def report(self):
+        return (self.name, self.total, self.failures)
+
+
+class ValueConstraints:
+    def __init__(self, constraints: List[ValueConstraint] = []):
+        self.constraints = constraints
+
+    @staticmethod
+    def from_protobuf(msg: ValueConstraintMsgs) -> 'ValueConstraints':
+        v = [ValueConstraint.from_proto(c) for c in msg.constraints]
+        if len(v) > 0:
+            return ValueConstraints(v)
+        return None
+
+    def to_protobuf(self) -> ValueConstraintMsgs:
+        v = [c.to_protobuf() for c in self.constraints]
+        if len(v) > 0:
+            vcmsg = ValueConstraintMsgs()
+            vcmsg.constraints.extend(v)
+            return vcmsg
+        return None
+
+    def add_constraint(self, c: ValueConstraint):
+        self.constraints.append(c)
+
+    def track(self, v):
```

Comment:
Same as above.

Also consider making an abstract class for these Constraints?
https://docs.python.org/3/library/abc.html

Reply:
we can push this to another PR, creating a issue is an ok reponse to resolve

Reply:
Please correct me if I'm wrong, but I don't think and ABC buys much here.
About the only things these two classes share completely is the report() method.
__init__, to_protobuf, from_protbuf, and update  take or produce different types so those would have to be defined in the subclasses anyway.  The ValueConstraint and SummaryConstrainty instances are kept separate and their update methods take different parameters.  Is there a value in the abstraction in that case?

Reply:
one thing it adds, is check to make sure people who create new constraints implements at least these two methods, if we add a `NotImplementedError` for example. We need to refactors the to/from protobufs too prob in a separate abstr class, or potential refactor into a visitor or another pattern.

Reply:
but again I feel we can do this later.

Reply:
ok let's get this in as-is and addredd ABC in another PR, if that's ok with Andy.